### PR TITLE
WIP - Feat oracle simplified

### PR DESCRIPTION
**Pull request by GalloDaSballo**
- [x] Tinfoil calls
- [x] Simple Primary Secondary Logic
- [x] Braindead Interface
- [X] Chainlink Caller Scaffolded
- [ ] Chainlink Caller Tested
- [x] Setup Tests for the TinfoilCalls (OOG, Burn all Gas, Return Bomb, Incorrect Return Values, Incorrect Type)
- [ ] Brutalize the TinfoilCalls (OOG, Burn all Gas, Return Bomb, Incorrect Return Values, Incorrect Type)
- [ ] Extra Caller
- [ ] Differential Fuzzing of CL Caller and Price Feed
- [ ] excessivelySafeCall to library

### CODE REVIEW

**Comment by dapp-whisperer** [comment]
# BraindeadFeed
Oracles return a single value, with 0 being flagged as INVALID_PRICE.

Determination of the _price value and it's validity_ is moved to the adapter level, which would also need to be implemented for our oracles.

To set an oracle address, the price is called and must return a non-zero price.

### fetchPrice() logic
- get primary price
- if valid, store and return

- if primary invalid, and we have a backup oracle present, get backup price
- if valid, store and return

- return last good price at the end
  - situation 1: primary oracle only (no secondary), returns INVALID_PRICE
  - situation 2: primary oracle and secondary oracle return INVALID_PRICE

**Comment by dapp-whisperer** [comment]
# Price Validity
What would we potentially want to consider in each oracle adapter?

- freshness of price value
- value versus previous value from that oracle (deviation threshold)
- value versus another reference oracle
- value versus lastGoodPrice

## Existing Chainlink Logic
Get the current and previous CL rounds. If there is no previous CL round, we'll have issues.
`_getCurrentChainlinkResponse`
`_getPrevChainlinkResponse`

CL is considered "broken" if either the current or previous responses are considered "bad"

a response is considered "bad" if:
- the success flag is not true
- either one of the component oracles has a timestamp of zero
- either one of the component oracles has a timestamp in the future according to block.timestamp

File: packages/contracts/contracts/BraindeadFeed.sol [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        uint256 currentPrice = IOracleCaller(_newPrimary).getLatestPrice();
+        require(currentPrice != 0, "Primary Oracle Must Work");
+
+        primaryOracle = _newPrimary;
+
+    }
+
+    function setSecondaryOracle(address _newSecondary) external requiresAuth {
+        uint256 currentPrice = IOracleCaller(_newSecondary).getLatestPrice();
+        require(currentPrice != 0, "Primary Oracle Must Work");
```
Comment (rayeaster) [inline packages/contracts/contracts/BraindeadFeed.sol:66]: minor revert typo. should be "BraindeadFeed: Secondary Oracle Must Work!"

File: packages/contracts/contracts/BraindeadFeed.sol [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        lastGoodPrice = _currentPrice;
+        emit LastGoodPriceUpdated(_currentPrice);
+    }
+
+    // Tinfoil Mode
+    // Give up to 2_M gas
+
+    function tinfoilCall(address _target, bytes memory _calldata) public returns (uint256) {
+        // Cap gas at 2 MLN, we don't care about 1/64 cause we expect oracles to consume way less than 200k gas
+        uint256 cappedGas = gasleft() > 2_000_000 ? 2_000_000 : gasleft();
```
Comment (rayeaster) [inline packages/contracts/contracts/BraindeadFeed.sol:208]: move this magic 2_000_000 to a constant variable and cache `gasleft()`?

File: packages/contracts/contracts/BraindeadFeed.sol [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+contract BraindeadFeed is IPriceFeed, AuthNoOwner {
+    string public constant NAME = "PriceFeed";
+
+    // The last good price seen from an oracle by Liquity
+    uint256 public lastGoodPrice;
+
+    address public primaryOracle;
+    address public secondaryOracle;
+
+    uint256 INVALID_PRICE = 0;
```
Comment (rayeaster) [inline packages/contracts/contracts/BraindeadFeed.sol:28]: add `constant`?

File: packages/contracts/contracts/BraindeadFeed.sol [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            uint256 secondaryOraclePrice = IOracleCaller(_secondaryOracle).getLatestPrice();
+
+            if(secondaryOraclePrice != 0) {
+                secondaryOracle = _secondaryOracle;
+            }
+        }
+    }
+
+    function setPrimaryOracle(address _newPrimary) external requiresAuth {
+        uint256 currentPrice = IOracleCaller(_newPrimary).getLatestPrice();
```
Comment (rayeaster) [inline packages/contracts/contracts/BraindeadFeed.sol:57]: do we need to use `tinfoilCall()` to `_newPrimary` here instead calling directly?

File: packages/contracts/contracts/Interfaces/IOracleCaller.sol [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.17;
+
+import "./IPool.sol";
```
Comment (rayeaster) [inline packages/contracts/contracts/Interfaces/IOracleCaller.sol:5]: seems this import `IPool` not used?

**Comment by rayeaster** [comment]
It looks like all "invalid" scenarios will come down to a simple "zero" price returned for both primary and secondary oracles. It is now the low-level adapter's responsibility to guarantee price [validity](https://github.com/ebtc-protocol/ebtc/pull/723#issuecomment-1821716565) checks

File: packages/contracts/contracts/BraindeadFeed.sol (hunk 4 above)
Comment (GalloDaSballo) [reply packages/contracts/contracts/BraindeadFeed.sol:57]: No because if the call reverts normally, the oracle should not be added

**Comment by GalloDaSballo** [comment]
@rayeaster rest of comments are 10 mins, please just commit them

**Comment by GalloDaSballo** [comment]
Setup tests for Tinfoil Call

Also sent early draft of CL Feed changed

We could also add a Min and Max Check in the CL Caller (QA report we received which is somewhat legitimate although very edge case), see: https://etherscan.io/address/0xdBe1941BFbe4410D6865b9b7078e0b49af144D2d#readContract

**Comment by dapp-whisperer** [comment]
do we need fallbackCall and authority address in the `ChainlinkCaller`?

```
constructor(
        address _fallbackCallerAddress,
        address _authorityAddress,
        address _collEthCLFeed,
        address _ethBtcCLFeed
    ) {
```

**Comment by dapp-whisperer** [comment]
# (Dapp) Testing Notes
## Governance / access control tests
- authorized users can modify primary oracle
- authorized users can modify secondary oracle
- a primary oracle that reports valid price when set as oracle can be set
- a primary oracle that does not report valid price when set as oracle reverts
- a secondary oracle that reports valid price when set as oracle can be set
- a secondary oracle that does not report valid price when set as oracle reverts

## State machine tests
### primary (CL) only
every oracle should be tested in isolation to ensure all it's conditions work
- oracle is broken under each broken condition and returns INVALID_PRICE
- oracle is frozen under each frozen condition and returns INVALID_PRICE
- oracle is above max deviation under max deviation condition and returns INVALID_PRICE
  - (confirm the thresholds for both upwards and downwards deviation)

### primary + mock secondary oracle


- [ ] tests to switchout both oracles and expected results
- [ ] add everything to deploy script
